kpi/deployment_backends/kobocat_backend.py

```diff
@@ -951,6 +961,154 @@ def __get_submissions_in_xml(self, **params):
+
+            self._kobocat_request('POST',
+                                  url=metadata_url,
+                                  sync_media_files=True,
+                                  **kwargs)
+
+        # Process deleted files in case two entries contain the same file but
+        # one is flagged as deleted
+        asset_files = self.asset.asset_files.filter(
+            file_type=AssetFile.FORM_MEDIA
+        ).order_by('date_deleted')
+
+        url = self.external_to_internal_url(self.backend_response['url'])
+        response = self._kobocat_request('GET', url)
+        kc_files = defaultdict(dict)
+        for metadata in response.get('metadata', []):
+            if metadata['data_type'] == 'media':
+
+                kc_files[metadata['data_value']] = {
+                    'url': metadata['url'],
+                    'md5': metadata['file_hash'],
+                    'from_kpi': metadata['from_kpi'],
+                }
+
+        kc_filenames = kc_files.keys()
+        for file in asset_files:
+            uniq = (
+                file.metadata['filename']
+                if not file.is_remote_url
+                else file.metadata['redirect_url']
+            )
+
+            # File does not exist in KC
+            if uniq not in kc_filenames:
+                if file.date_deleted is None:
+                    # New file
+                    _upload_to_kc(file)
+                else:
+                    # Orphan, delete it
+                    file.delete(force=True)
+                continue
+
+            # Existing file
+            if uniq in kc_filenames:
+                kc_file = kc_files[uniq]
+                if file.date_deleted is None:
+                    # If md5 differs, we need to re-upload it.
+                    if file.metadata.get('hash') != kc_file['md5']:
+                        _delete_kc_file(kc_file)
+                        _upload_to_kc(file)
+                elif kc_file['from_kpi']:
+                    _delete_kc_file(kc_file, file)
+                else:
+                    # Remote file has been uploaded directly to KC. We
+                    # cannot delete it, but we need to vacuum KPI.
+                    file.delete(force=True)
+                    # Skip deletion of key corresponding to `uniq` in `kc_files`
+                    # to avoid unique constraint failure in case user deleted
+                    # and re-uploaded the same file in a row between
+                    # two deployments
+                    # Example:
+                    # - User uploads file1.jpg (pk == 1)
+                    # - User deletes file1.jpg (pk == 1)
+                    # - User re-uploads file1.jpg (pk == 2)
+                    # Next time, 'file1.jpg' is encountered in this loop,
+                    # it would try to re-upload to KC if its hash differs
+                    # from KC version and would fail because 'file1.jpg'
+                    # already exists in KC db.
```

Comment:
i don't quite understand. why would `file1.jpg (pk == 1)` be uploaded since there's a check for `file.date_deleted is None` before `_upload_to_kc()`?

Reply:
asset a123 (kpi) -> xform x123 (kobocat)

* file1.jpg already associated with a123 **and** already pushed to x123
* user deletes file1.jpg from a123
* user uploads new file1.jpg to a123 (different hash)
* this loop tries to push new file1.jpg to x123 but fails because of non-unique filename